bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/handler/BroadlinkRemoteHandler.java

```diff
@@ -0,0 +1,496 @@
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.broadlink.internal.BroadlinkBindingConstants;
+import org.openhab.binding.broadlink.internal.BroadlinkBindingConstants.CodeType;
+import org.openhab.binding.broadlink.internal.BroadlinkMappingService;
+import org.openhab.binding.broadlink.internal.BroadlinkRemoteDynamicCommandDescriptionProvider;
+import org.openhab.binding.broadlink.internal.Utils;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.storage.StorageService;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.openhab.core.util.HexUtils;
+
+/**
+ * Remote blaster handler superclass
+ *
+ * @author Cato Sognen - Initial contribution
+ * @author John Marshall - V3 rewrite with dynamic command description provider
+ */
+@NonNullByDefault
+public abstract class BroadlinkRemoteHandler extends BroadlinkBaseThingHandler {
+    public static final byte COMMAND_BYTE_SEND_CODE = 0x02;
+
+    // IR commands
+    public static final byte COMMAND_BYTE_ENTER_LEARNING = 0x03;
+    public static final byte COMMAND_BYTE_CHECK_LEARNT_DATA = 0x04;
+
+    // RF commands
+    public static final byte COMMAND_BYTE_ENTER_RF_FREQ_LEARNING = 0x19; // Sweep frequency
+    public static final byte COMMAND_BYTE_CHECK_RF_FREQ_LEARNING = 0x1A; // Check frequency
+    public static final byte COMMAND_BYTE_EXIT_RF_FREQ_LEARNING = 0x1E; // Cancel sweep frequency
+    public static final byte COMMAND_BYTE_FIND_RF_PACKET = 0x1B; // Find RF packet
+    public static final byte COMMAND_BYTE_CHECK_RF_DATA = 0x4; // Check data
+
+    private final BroadlinkRemoteDynamicCommandDescriptionProvider commandDescriptionProvider;
+    private final StorageService storageService;
+    protected @Nullable BroadlinkMappingService mappingService;
+
+    public BroadlinkRemoteHandler(Thing thing,
+            BroadlinkRemoteDynamicCommandDescriptionProvider commandDescriptionProvider,
+            StorageService storageService) {
+        super(thing);
+        this.commandDescriptionProvider = commandDescriptionProvider;
+        this.storageService = storageService;
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+        this.mappingService = new BroadlinkMappingService(commandDescriptionProvider,
+                new ChannelUID(thing.getUID(), BroadlinkBindingConstants.COMMAND_CHANNEL),
+                new ChannelUID(thing.getUID(), BroadlinkBindingConstants.RF_COMMAND_CHANNEL), this.storageService);
+    }
+
+    @Override
+    public void dispose() {
+        if (this.mappingService != null) {
+            this.mappingService.dispose();
```

Comment:
Please fix all compiler warnings.
```suggestion
        BroadlinkMappingService mappingService = this.mappingService;
        if (mappingService != null) {
            mappingService.dispose();
            this.mappingSerivce = null;
```